torch/distributed/_state_dict_utils.py

```diff
@@ -457,3 +457,93 @@ def tensor_func(
         return False
 
     return True
+
+
+class _TensorInfo(NamedTuple):
+    size: torch.Size
+    dtype: torch.dtype
+
+
+def _broadcast_tensors(
+    full_state_dict: Dict[str, Any],
+    local_state_dict: Dict[str, Any],
+    keys: List[str],
+    device: torch.device,
+    pg: Optional[dist.ProcessGroup] = None,
+) -> None:
+    tensors = []
+    for key in keys:
+        if dist.get_rank() == 0:
+            full_state = full_state_dict[key]
+            assert isinstance(full_state, torch.Tensor)
+            full_tensor = full_state.detach().to(device)
+        else:
+            tensor_info = full_state_dict[key]
+            full_tensor = torch.empty(
+                size=tensor_info.size,
+                device=device,
+                dtype=tensor_info.dtype,
+            )
+
+        tensors.append(full_tensor)
+        local_state = local_state_dict.get(key, None)
+        if local_state is None:
+            continue
+        elif isinstance(local_state, DTensor):
+            local_state_dict[key] = (local_state, full_tensor)
+        else:
+            local_state_dict[key] = full_tensor
+
+    if pg is None:
+        pg = dist.distributed_c10d._get_default_group()
+    dist._broadcast_coalesced(pg, tensors, 500, 0)
+
```

Comment:
Another detail: I think we should be careful to use `dist._broadcast_coalesced` today since it will call `recordStream` on the input tensors. I think to avoid this, we can either manually use the coalescing context from Python or just use `dist.broadcast`.

Reply:
Our current coalescing context does not support broadcast. I can take a look to see how can we can the support. But even with `recordStream`, is that really going to cause problem if we only do this when loading the checkpoint?

Reply:
we had a debug last week in torchtune around this. for 70b, recordstream during loading checkpoint peaks memeory to 78G. it’s reduced to 22G without recordstream

Reply:
we have time to figure out coalescing ctx though 

the above comment of avoiding full sd in GPU is critical and time sensitive

we were doing broadcast+distribute_tensor together in one for loop in torchtune. maybe it also apply here

Reply:
oh, right. This implementation may blow out the GPU memory. I can change it to broadcast every 10 tensors. Or using _broadcast_coalesced is still not acceptable even if I synchronize after the broadcast?

Reply:
@fegin I missed this. User need to set `TORCH_NCCL_AVOID_RECORD_STREAMS` when using `dist._broadcast_coalesced` to release GPU memory timely. Maybe regular broadcast to avoid recordstreams before we had a solution for coalescing?